Doc: Add blogpost for honor PV reclaim policy fix (#30556)
kubernetes/website: content/en/blog/_posts/2021-11-18-honor-pv-reclaim-policy.md

---
layout: blog

> title: "Kubernetes 1.23 Feature: Honor PV Reclaim Policy"

Reviewer: ```suggestion
title: "Kubernetes 1.23: PersistentVolume reclaim improvements"
```
?
(and a new slug to match)

date: 2021-11-18T10:00:00-08:00
slug: kubernetes-1-23-feature-honor-pv-reclaim-policy
---

> **Authors:** Deepak Kinni, VMware

Reviewer: ```suggestion
**Author:** Deepak Kinni (VMware)
```

> [Persistent Volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) are associated with [Reclaim Policy](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaim-policy). The Reclaim Policy is used to determine the actions that need to be taken by the storage backend on deletion of the PV. In case where the Reclaim Policy is `Delete` the expectation is that the storage backend releases the storage resource that was allocated for the PV. In essence, the Reclaim policy needs to honored on PV deletion.

Reviewer: ```suggestion
[PersistentVolumes](/docs/concepts/storage/persistent-volumes/) (or PVs for short) are
associated with a [reclaim policy](/docs/concepts/storage/persistent-volumes/#reclaim-policy).
The reclaim policy is used to determine the actions that need to be taken by the storage
backend on deletion of the PV.
Where the reclaim policy is `Delete`, the expectation is that the storage backend
releases the storage resource that was allocated for the PV. In essence, the reclaim
policy needs to honored on PV deletion.

With the recent Kubernetes v1.23 release, an alpha feature lets you configure your
cluster to behave that way and honor the configured reclaim policy.
```
?

> ## What is the existing behavior?

Reviewer: ```suggestion
## How did reclaim work in previous Kubernetes releases?
```

> Normally, if the volume is to be deleted, then the expectation is to delete the PVC for a bound PV-PVC pair. However, there is no restriction to delete a PV prior to deleting a PVC.

Reviewer: We are using “PVC” here without first introducing the term. Some readers will be puzzled, and we should help them with a (brief) summary of how PVCs and PVs are related, plus a hyperlink for further reading.

> The steps below demonstrate the current behavior

Reviewer: ```suggestion
First, I'll demonstrate the behavior for clusters that are running an older version of Kubernetes.
```

#### Create PVC and wait for it to get bound.

```shell
# kubectl create -f pvc.yaml 
persistentvolumeclaim/example-vanilla-block-pvc created

> ```

Reviewer: This isn't a valid shell script. See https://kubernetes.io/docs/contribute/style/style-guide/#code-snippet-formatting for our style guide recommendations.

```shell
# kubectl get pvc
NAME                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS               AGE
example-vanilla-block-pvc   Bound    pvc-6791fdd4-5fad-438e-a7fb-16410363e3da   5Gi        RWO            example-vanilla-block-sc   19s
```

> #### Delete PV.

Reviewer: nit: Omit the period here

> It can be observed that deleting a bound PV will not return back control to the shell.

Reviewer: ```suggestion
When I try to delete a bound PV, the cluster blocks and the `kubectl` tool does
not return back control to the shell; for example:
```

pvc-6791fdd4-5fad-438e-a7fb-16410363e3da   5Gi        RWO            Delete           Terminating   default/example-vanilla-block-pvc   example-vanilla-block-sc            2m23s
```

#### Delete PVC

```shell
# kubectl delete pvc example-vanilla-block-pvc
persistentvolumeclaim "example-vanilla-block-pvc" deleted
```

The PV object from the cluster also gets deleted

```shell
# kubectl get pv pvc-6791fdd4-5fad-438e-a7fb-16410363e3da
Error from server (NotFound): persistentvolumes "pvc-6791fdd4-5fad-438e-a7fb-16410363e3da" not found
```

However, the underlying storage resource is not deleted, and needs to be removed manually.

To sum it up, Reclaim policy associated with the Persistent Volume is currently ignored under certain circumstance. For a `Bound` PV-PVC pair the ordering of PV-PVC deletion determines whether the PV delete reclaim policy is honored. The PV honors the reclaim policy if the PVC is deleted, however, if the PV is deleted prior to deleting the PVC then the Reclaim policy is not exercised. As a result of this behavior, the associated storage asset in the external infrastructure is not removed.

> ## What is the new behavior?

Reviewer: ```suggestion
## Volume reclaim with Kubernetes v1.23 {#new-behavior}
```

The new behavior ensures that the underlying storage object is deleted from the backend when users attempt to delete a PV manually.

> The fix is introduced as an alpha in 1.23 under the feature-gate `HonorPVReclaimPolicy`. Update the `external-provisioner` to the `v4.0.0` and enable the feature gate.

Reviewer: **This (the fact it's an alpha feature and off by default) needs highlighting really prominently.**

#### How does it work?

The new behavior is achieved by adding a finalizer `external-provisioner.volume.kubernetes.io/finalizer` on new and existing PVs, the finalizer is only removed after the storage from backend is deleted.

An example of a PV with the finalizer, notice the new finalizer in the finalizers list

> ```shell

Reviewer: I'd mark this block as YAML (and put the command separately).
